packages/typescript-estree/src/create-program/createProjectProgram.ts

```diff
@@ -34,65 +34,82 @@ function createProjectProgram(
 ): ASTAndProgram | undefined {
   log('Creating project program for: %s', extra.filePath);
 
-  const astAndProgram = firstDefined(
-    getProgramsForProjects(code, extra.filePath, extra),
-    currentProgram => getAstFromProgram(currentProgram, extra),
+  const programsForProjects = getProgramsForProjects(
+    code,
+    extra.filePath,
+    extra,
+  );
+  const astAndProgram = firstDefined(programsForProjects, currentProgram =>
+    getAstFromProgram(currentProgram, extra),
   );
 
-  if (!astAndProgram && !createDefaultProgram) {
-    // the file was either not matched within the tsconfig, or the extension wasn't expected
-    const errorLines = [
-      '"parserOptions.project" has been set for @typescript-eslint/parser.',
-      `The file does not match your project config: ${path.relative(
-        extra.tsconfigRootDir || process.cwd(),
-        extra.filePath,
-      )}.`,
-    ];
-    let hasMatchedAnError = false;
+  // The file was either matched within the tsconfig, or we allow creating a default program
+  if (astAndProgram || createDefaultProgram) {
+    return astAndProgram;
+  }
 
-    const extraFileExtensions = extra.extraFileExtensions || [];
+  const describeFilePath = (filePath: string): string =>
+    path.relative(extra.tsconfigRootDir || process.cwd(), filePath);
```

Comment:
I wonder if we should do like this to make it clearer?

```suggestion
  const describeFilePath = (filePath: string): string => {
    const relative = path.relative(extra.tsconfigRootDir || process.cwd(), filePath);
    if (extra.tsconfigRootDir) {
      return `<tsconfigRootDir>/${relative}`;
    }
    return `<cwd>/${relative}`;
  };
```